app/models/report.rb

```diff
@@ -82,35 +96,46 @@ def self.expire(conditions, batch_size, sleep_time)
     cond = "created_at < \'#{created}\'"
     cond += " and status = #{status}" unless status.nil?
     total_count = 0
-    report_ids = []
-    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    # find the first (oldest) report to be deleted
+    report_id_max = where(cond).reorder('').maximum(:id)
+    return unless report_id_max
+    # delete log entries in batches
     loop do
+      deleted = nil
       Report.transaction do
-        batch_start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
-        report_ids = where(cond).reorder('').limit(batch_size).pluck(:id)
-        if report_ids.count > 0
-          log_count = Log.unscoped.where(:report_id => report_ids).reorder('').delete_all
-          count = where(:id => report_ids).reorder('').delete_all
-          total_count += count
-          rate = (count / (Process.clock_gettime(Process::CLOCK_MONOTONIC) - batch_start_time)).to_i
-          Foreman::Logging.with_fields(expired_logs: log_count, expired_total: count, expire_rate: rate) do
-            logger.info "Expired #{count} reports and #{log_count} logs at rate #{rate} reports/sec"
-          end
+        with_logging("logs") do
+          report_id_min = report_id_max - batch_size
```

Comment:
What if there is a different report type in between the two? e.g., let's imagine batch size of 3 just for the example:

id | type
---|-----
1 | ConfigReport
2 | ArfReport
3 | ConfigReport

This will cause the ARF report and it's logs to be expired as well.

Also, looking at the table now looks like there's an off-by-one error here - `report_id_min` will be 0, not 1

Reply:
User can only configure a date from reports should be expired.  Let's say it's ID=3, then thank to monotonic sequences in postgresql record with ID=2 is guaranteed to be created before report ID=3, thus it should have been deleted.

Can you explain the off-by-one error in more detail? ;-)